File: desktop/src/main/marketplace-file-reader.ts
// Reads the raw markdown for a plugin's SKILL.md / command / agent file so
// the in-app file viewer can render it. Installed plugins resolve to the
// on-disk copy; non-installed fall back to a raw GitHub URL derived from the
// marketplace entry's sourceType/sourceRef.
//
// We glob for the file instead of assuming a fixed layout — youcoded-core lays
// skills out under core/skills/, life/skills/, productivity/skills/, while
// single-layer plugins like civic-report use skills/<name>/SKILL.md flat.

import fs from 'fs';
import path from 'path';
import os from 'os';
import type { SkillEntry } from '../shared/types';
import { YOUCODED_PLUGINS_DIR } from './claude-code-registry';

export type ComponentKind = 'skill' | 'command' | 'agent';

export interface ReadComponentArgs {
  pluginId: string;
  kind: ComponentKind;
  name: string;
}

export interface ReadComponentResult {
  content: string;
  source: 'local' | 'remote';
  path: string; // for display / debugging only
}

const CLAUDE_PLUGINS_ROOT = path.join(os.homedir(), '.claude', 'plugins');
const REGISTRY_BASE = `https://raw.githubusercontent.com/itsdestin/wecoded-marketplace/${process.env.YOUCODED_MARKETPLACE_BRANCH || 'master'}`;

function resolvePluginDir(id: string): string | null {
  // Core toolkit lives at ~/.claude/plugins/youcoded-core (not the marketplace
  // subtree); marketplace-installed plugins live under YOUCODED_PLUGINS_DIR.
  const topLevel = path.join(CLAUDE_PLUGINS_ROOT, id);
  if (fs.existsSync(topLevel)) return topLevel;
  const marketplace = path.join(YOUCODED_PLUGINS_DIR, id);
  if (fs.existsSync(marketplace)) return marketplace;
  return null;
}

// Relative path shape for each component kind in a plugin tree.
function relativePathsFor(kind: ComponentKind, name: string): string[] {
  if (kind === 'skill') return [`skills/${name}/SKILL.md`];
  if (kind === 'command') return [`commands/${name}.md`];
  return [`agents/${name}.md`];
}

// Walk up to `maxDepth` levels looking for any candidate relative path.
// Depth 4 covers youcoded-core's `core/skills/...` / `life/skills/...` /
// `productivity/skills/...` layouts without descending into node_modules
// or other deep trees.
function findLocalFile(rootDir: string, relative: string, maxDepth = 4): string | null {
  const direct = path.join(rootDir, relative);
  if (fs.existsSync(direct)) return direct;

  const queue: Array<{ dir: string; depth: number }> = [{ dir: rootDir, depth: 0 }];
  while (queue.length > 0) {
    const { dir, depth } = queue.shift()!;
    if (depth >= maxDepth) continue;
    let entries: fs.Dirent[];
    try { entries = fs.readdirSync(dir, { withFileTypes: true }); }
    catch { continue; }
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      if (entry.name.startsWith('.') || entry.name === 'node_modules') continue;
      const sub = path.join(dir, entry.name);
      const candidate = path.join(sub, relative);
      if (fs.existsSync(candidate)) return candidate;
      queue.push({ dir: sub, depth: depth + 1 });
    }
  }
  return null;
}

// Build raw.githubusercontent.com URLs for the given relative path. We return
// multiple candidates and try them in order — covers both flat layouts and
// youcoded-core's layered core/life/productivity prefixes.
function buildRemoteCandidates(entry: SkillEntry, relative: string): string[] {
  const prefixes = ['', 'core/', 'life/', 'productivity/'];

  if (entry.sourceType === 'local' && entry.sourceRef) {
    // Plugin lives at <marketplace-repo>/<sourceRef>/...
    return prefixes.map(p => `${REGISTRY_BASE}/${entry.sourceRef}/${p}${relative}`);
  }

  if ((entry.sourceType === 'url' || entry.sourceType === 'git-subdir') && entry.sourceRef) {
    const parsed = parseGithubRepo(entry.sourceRef);
    if (!parsed) return [];
    const subdir = entry.sourceType === 'git-subdir' && entry.sourceSubdir
      ? `${entry.sourceSubdir.replace(/\/$/, '')}/` : '';
    const base = `https://raw.githubusercontent.com/${parsed.owner}/${parsed.repo}/${parsed.branch}`;
    return prefixes.map(p => `${base}/${subdir}${p}${relative}`);
  }

  return [];
}

function parseGithubRepo(url: string): { owner: string; repo: string; branch: string } | null {
  // Accept https://github.com/<owner>/<repo>(.git)(#branch)
  const m = url.match(/github\.com\/([^/]+)\/([^/#]+?)(?:\.git)?(?:#(.+))?$/);
  if (!m) return null;
  return { owner: m[1], repo: m[2], branch: m[3] || 'master' };
}

export async function readComponent(
  args: ReadComponentArgs,
  fetchIndex: () => Promise<SkillEntry[]>,
): Promise<ReadComponentResult> {
  const { pluginId, kind, name } = args;
  const relatives = relativePathsFor(kind, name);

  // Local first — cheap and works offline.
  const installDir = resolvePluginDir(pluginId);
  if (installDir) {
    for (const rel of relatives) {
      const hit = findLocalFile(installDir, rel);
      if (hit) {
        const content = fs.readFileSync(hit, 'utf8');
        return { content, source: 'local', path: hit };
      }
    }
  }

  // Remote fallback — need the registry entry for sourceType/sourceRef.
  const index = await fetchIndex();
  const entry = index.find(e => e.id === pluginId);
  if (!entry) throw new Error(`Plugin not found in marketplace: ${pluginId}`);

  for (const rel of relatives) {
    const candidates = buildRemoteCandidates(entry, rel);
    for (const url of candidates) {
      try {
        const res = await fetch(url);
        if (res.ok) {
          const content = await res.text();
          return { content, source: 'remote', path: url };
        }
      } catch { /* try next */ }
    }
  }

  throw new Error(`File not found: ${kind} "${name}" in ${pluginId}`);
}

File: app/src/main/kotlin/com/youcoded/app/skills/MarketplaceFileReader.kt
package com.youcoded.app.skills

import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.URL

/**
 * Kotlin port of desktop's marketplace-file-reader.ts. Reads raw markdown for a
 * plugin's SKILL.md / command / agent file so the in-app file viewer can render
 * it. Tries the on-disk install first, then falls back to a raw GitHub URL
 * derived from the marketplace entry's sourceType/sourceRef.
 *
 * We glob instead of assuming a fixed layout — youcoded-core lays skills out
 * under core/skills/, life/skills/, productivity/skills/, while single-layer
 * plugins use skills/<name>/SKILL.md flat.
 */
object MarketplaceFileReader {
    private const val REGISTRY_BASE =
        "https://raw.githubusercontent.com/itsdestin/wecoded-marketplace/master"

    /**
     * Returns `{content, source, path}` on success or `{error}` on failure —
     * shape matches desktop's readComponent() exactly so the renderer can
     * treat both platforms identically.
     */
    fun readComponent(
        homeDir: File,
        pluginId: String,
        kind: String,
        name: String,
        marketplaceIndex: JSONArray,
    ): JSONObject {
        val relatives = relativePathsFor(kind, name)

        // Local first — cheap and works offline.
        val installDir = resolvePluginDir(homeDir, pluginId)
        if (installDir != null) {
            for (rel in relatives) {
                val hit = findLocalFile(installDir, rel)
                if (hit != null) {
                    return JSONObject().apply {
                        put("content", hit.readText())
                        put("source", "local")
                        put("path", hit.absolutePath)
                    }
                }
            }
        }

        // Remote fallback — need the registry entry for sourceType/sourceRef.
        val entry = findEntry(marketplaceIndex, pluginId)
            ?: return JSONObject().put("error", "Plugin not found in marketplace: $pluginId")

        for (rel in relatives) {
            for (url in buildRemoteCandidates(entry, rel)) {
                val content = fetchText(url)
                if (content != null) {
                    return JSONObject().apply {
                        put("content", content)
                        put("source", "remote")
                        put("path", url)
                    }
                }
            }
        }
        return JSONObject().put("error", "File not found: $kind \"$name\" in $pluginId")
    }

    // Core toolkit lives at ~/.claude/plugins/<id>/ (cloned by install.sh);
    // marketplace-installed plugins live under the marketplaces/ subtree.
    private fun resolvePluginDir(homeDir: File, id: String): File? {
        val root = File(homeDir, ".claude/plugins")
        val topLevel = File(root, id)
        if (topLevel.exists()) return topLevel
        val marketplace = File(root, "marketplaces/youcoded/plugins/$id")
        if (marketplace.exists()) return marketplace
        return null
    }

    private fun relativePathsFor(kind: String, name: String): List<String> = when (kind) {
        "skill" -> listOf("skills/$name/SKILL.md")
        "command" -> listOf("commands/$name.md")
        else -> listOf("agents/$name.md")
    }

    // BFS up to maxDepth=4, which covers youcoded-core's
    // core/skills/, life/skills/, productivity/skills/ layouts without
    // descending into node_modules or hidden dirs.
    private fun findLocalFile(rootDir: File, relative: String, maxDepth: Int = 4): File? {
        val direct = File(rootDir, relative)
        if (direct.exists()) return direct

        data class Node(val dir: File, val depth: Int)
        val queue = ArrayDeque<Node>().apply { add(Node(rootDir, 0)) }
        while (queue.isNotEmpty()) {
            val node = queue.removeFirst()
            if (node.depth >= maxDepth) continue
            val children = try { node.dir.listFiles() ?: emptyArray() } catch (_: Exception) { emptyArray() }
            for (child in children) {
                if (!child.isDirectory) continue
                if (child.name.startsWith(".") || child.name == "node_modules") continue
                val candidate = File(child, relative)
                if (candidate.exists()) return candidate
                queue.add(Node(child, node.depth + 1))
            }
        }
        return null
    }

    private fun findEntry(index: JSONArray, pluginId: String): JSONObject? {
        for (i in 0 until index.length()) {
            val entry = index.optJSONObject(i) ?: continue
            if (entry.optString("id") == pluginId) return entry
        }
        return null
    }

    // Builds raw.githubusercontent.com URL candidates covering both flat
    // layouts and youcoded-core's layered core/life/productivity prefixes.
    private fun buildRemoteCandidates(entry: JSONObject, relative: String): List<String> {
        val prefixes = listOf("", "core/", "life/", "productivity/")
        val sourceType = entry.optString("sourceType")
        val sourceRef = entry.optString("sourceRef")
        if (sourceRef.isEmpty()) return emptyList()

        if (sourceType == "local") {
            // Plugin lives at <marketplace-repo>/<sourceRef>/...
            return prefixes.map { p -> "$REGISTRY_BASE/$sourceRef/$p$relative" }
        }

        if (sourceType == "url" || sourceType == "git-subdir") {
            val parsed = parseGithubRepo(sourceRef) ?: return emptyList()
            val (owner, repo, branch) = parsed
            val subdir = if (sourceType == "git-subdir") {
                val raw = entry.optString("sourceSubdir").trimEnd('/')
                if (raw.isEmpty()) "" else "$raw/"
            } else ""
            val base = "https://raw.githubusercontent.com/$owner/$repo/$branch"
            return prefixes.map { p -> "$base/$subdir$p$relative" }
        }
        return emptyList()
    }

    private fun parseGithubRepo(url: String): Triple<String, String, String>? {
        val m = Regex("""github\.com/([^/]+)/([^/#]+?)(?:\.git)?(?:#(.+))?$""").find(url) ?: return null
        val branch = m.groupValues[3].ifEmpty { "master" }
        return Triple(m.groupValues[1], m.groupValues[2], branch)
    }

    // Returns body on HTTP 200, null on any failure (404, network, etc).
    // URL().readText() throws on non-2xx, so any exception means "try next".
    private fun fetchText(url: String): String? = try {
        URL(url).readText()
    } catch (_: Exception) {
        null
    }
}
